KSP task (non-)registration happens too soon

I have a gradle plugin that tries to conditionally enable KSP and add relevant KSP processors to it. However, I found that KSP's task registration can be unintentionally deleted by a simple `project.afterEvaluate` in gradle, resulting in the task never being registered and thus not run.

Example

```kotlin
// in build.gradle.kts, or a gradle plugin

project.afterEvaluate {
  if (enableKsp) {
    pluginManager.apply("com.google.devtools.ksp")

    dependencies.add("ksp", "some.ksp:processor:1.0.0")
  }
}
```

will result in KSP being applied but no `ksp*Kotlin` task ever being registered and thus never running. If I add any dependency to the `ksp` configuration (even one without any KSP processors in it), it works and picks up the later-added dependency too.

I think I would expect this to always register the KSP task, then conditionally no-op at task-action-time rather than try to conditionally avoid adding the task.

This has come to a head in AGP 8.10.0, which now fully blocks KSP from being applied in an afterEvaluate, more or less preventing conditionally applying it in a convention plugin

```
Caused by: java.lang.RuntimeException: It is too late to call addKspConfigurations as the callbacks already executed.	
at com.android.build.api.extension.impl.OperationsRegistrar.addInternalOperation(OperationsRegistrar.kt:82)	
at com.android.build.api.extension.impl.OperationsRegistrar.addInternalOperation$default(OperationsRegistrar.kt:76)	
at com.android.build.api.extension.impl.AndroidComponentsExtensionImpl.addSourceSetConfigurations(AndroidComponentsExtensionImpl.kt:353)	
at com.android.build.api.extension.impl.AndroidComponentsExtensionImpl.addKspConfigurations(AndroidComponentsExtensionImpl.kt:148)	
at com.google.devtools.ksp.gradle.KspConfigurations$3.invoke(KspConfigurations.kt:108)	
at com.google.devtools.ksp.gradle.KspConfigurations$3.invoke(KspConfigurations.kt:104)	
at com.google.devtools.ksp.gradle.KspConfigurations._init_$lambda$5(KspConfigurations.kt:104)	
at org.gradle.api.internal.plugins.DefaultPluginManager$2.execute(DefaultPluginManager.java:261)	
at org.gradle.api.internal.plugins.DefaultPluginManager$2.execute(DefaultPluginManager.java:258)	
at org.gradle.internal.code.DefaultUserCodeApplicationContext$CurrentApplication$1.execute(DefaultUserCodeApplicationContext.java:124)	
at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction$1.run(DefaultCollectionCallbackActionDecorator.java:110)	
at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:30)	
at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:27)	
at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:67)	
at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:60)	
at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:167)	
at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:60)	
at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:48)	
at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction.execute(DefaultCollectionCallbackActionDecorator.java:107)	
at org.gradle.api.internal.DefaultDomainObjectCollection.all(DefaultDomainObjectCollection.java:185)	
at org.gradle.api.internal.plugins.DefaultPluginManager.withPlugin(DefaultPluginManager.java:264)	
at com.google.devtools.ksp.gradle.KspConfigurations.<init>(KspConfigurations.kt:104)	
at com.google.devtools.ksp.gradle.KspGradleSubplugin.apply(KspSubplugin.kt:220)	
at com.google.devtools.ksp.gradle.KspGradleSubplugin.apply(KspSubplugin.kt:80)	
at org.gradle.api.internal.plugins.ImperativeOnlyPluginTarget.applyImperative(ImperativeOnlyPluginTarget.java:55)	
at org.gradle.api.internal.plugins.RuleBasedPluginTarget.applyImperative(RuleBasedPluginTarget.java:51)	
at org.gradle.api.internal.plugins.DefaultPluginManager.addPlugin(DefaultPluginManager.java:190)	
at org.gradle.api.internal.plugins.DefaultPluginManager.access$100(DefaultPluginManager.java:54)	
at org.gradle.api.internal.plugins.DefaultPluginManager$AddPluginBuildOperation.run(DefaultPluginManager.java:285)	
at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:30)	
at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:27)	
at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:67)	
at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:60)	
at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:167)	
at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:60)	
at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:48)	
at org.gradle.api.internal.plugins.DefaultPluginManager.lambda$doApply$0(DefaultPluginManager.java:170)	
at org.gradle.internal.code.DefaultUserCodeApplicationContext.apply(DefaultUserCodeApplicationContext.java:44)	
at org.gradle.api.internal.plugins.DefaultPluginManager.doApply(DefaultPluginManager.java:169)	
at org.gradle.api.internal.plugins.DefaultPluginManager.apply(DefaultPluginManager.java:148)	
at foundry.gradle.FoundryExtension$applyTo$1.execute(FoundryExtension.kt:283)
```

--- a/gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspAATask.kt
+++ b/gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspAATask.kt
@@ -160,7 +160,11 @@ abstract class KspAATask @Inject constructor(
             ).apply {
                 isTransitive = false
             }
+            val incomingProcessors = processorClasspath.incoming.files
             val kspTaskProvider = project.tasks.register(kspTaskName, KspAATask::class.java) { kspAATask ->
+                kspAATask.onlyIf {
+                    !incomingProcessors.isEmpty()
+                }
                 kspAATask.kspClasspath.from(kspAADepCfg)
                 kspAATask.kspConfig.let { cfg ->
                     cfg.processorClasspath.from(processorClasspath)

--- a/gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspSubplugin.kt
+++ b/gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspSubplugin.kt
@@ -272,12 +272,7 @@ class KspGradleSubplugin @Inject internal constructor(private val registry: Tool
             val nonEmptyKspConfigurations =
                 kspConfigurations.find(kotlinCompilation)
                     .filter { it.allDependencies.isNotEmpty() }
-            if (nonEmptyKspConfigurations.isEmpty()) {
-                return project.provider { emptyList() }
-            }
             processorClasspath.extendsFrom(*nonEmptyKspConfigurations.toTypedArray())
-        } else if (processorClasspath.allDependencies.isEmpty()) {
-            return project.provider { emptyList() }
         }
 
         val target = kotlinCompilation.target.name
@@ -313,7 +308,11 @@ class KspGradleSubplugin @Inject internal constructor(private val registry: Tool
         )
 
         val kspCachesDir = getKspCachesDir(project, sourceSetName, target)
+        val incomingProcessors = processorClasspath.incoming.files
         fun configureAsKspTask(kspTask: KspTask, isIncremental: Boolean) {
+            kspTask.onlyIf {
+                !incomingProcessors.isEmpty()
+            }
             // depends on the processor; if the processor changes, it needs to be reprocessed.
             kspTask.dependsOn(processorClasspath.buildDependencies)
             kspTask.commandLineArgumentProviders.addAll(kspExtension.commandLineArgumentProviders)

--- a/gradle-plugin/src/test/kotlin/com/google/devtools/ksp/gradle/SourceSetConfigurationsTest.kt
+++ b/gradle-plugin/src/test/kotlin/com/google/devtools/ksp/gradle/SourceSetConfigurationsTest.kt
@@ -27,6 +27,7 @@ import com.google.devtools.ksp.processing.Resolver
 import com.google.devtools.ksp.processing.SymbolProcessor
 import com.google.devtools.ksp.symbol.KSAnnotated
 import com.google.devtools.ksp.symbol.KSClassDeclaration
+import org.gradle.testkit.runner.TaskOutcome
 import org.junit.Rule
 import org.junit.Test
 import org.junit.rules.TemporaryFolder
@@ -132,43 +133,35 @@ class SourceSetConfigurationsTest(val useKSP2: Boolean) {
     }
 
     @Test
-    fun configurationsForMultiplatformApp_doesNotCrossCompilationBoundaries() {
-        // Adding a ksp dependency on jvmParent should not leak into jvmChild compilation,
-        // even if the source sets depend on each other. This works because we use
-        // KotlinCompilation.kotlinSourceSets instead of KotlinCompilation.allKotlinSourceSets
+    fun configurationsForMultiplatformApp_skipEmptyKspTasks() {
         testRule.setupAppAsMultiplatformApp(
             """
                 kotlin {
-                    jvm("jvmParent") { }
-                    jvm("jvmChild") { }
+                    jvm { }
+                    js(IR) { browser() }
                 }
             """.trimIndent()
         )
         testRule.appModule.addMultiplatformSource("commonMain", "Foo.kt", "class Foo")
         testRule.appModule.buildFileAdditions.add(
             """
-                kotlin {
-                    sourceSets {
-                        this["jvmChildMain"].dependsOn(this["jvmParentMain"])
-                    }
-                }
                 dependencies {
-                    add("kspJvmParent", "androidx.room:room-compiler:2.4.2")
-                }
-                tasks.register("checkConfigurations") {
-                    doLast {
-                        // child has no dependencies, so task is not created.
-                        val parent = tasks.findByName("kspKotlinJvmParent")
-                        val child = tasks.findByName("kspKotlinJvmChild")
-                        require(parent != null)
-                        require(child == null)
-                    }
+                    add("kspJvm", "androidx.room:room-compiler:2.4.2")
                 }
             """.trimIndent()
         )
         testRule.runner()
-            .withArguments(":app:checkConfigurations")
-            .build()
+            .withArguments(":app:kspKotlinJvm", ":app:kspKotlinJs")
+            .build().let {
+                val kspKotlinJvm = it.task(":app:kspKotlinJvm")
+                val kspKotlinJs = it.task(":app:kspKotlinJs")
+                require(kspKotlinJvm != null)
+                require(kspKotlinJvm.outcome == TaskOutcome.SUCCESS)
+                // even though kspJs is not added, the task is created.
+                require(kspKotlinJs != null)
+                // kspKotlinJs has no dependencies, so task is skipped.
+                require(kspKotlinJs.outcome == TaskOutcome.SKIPPED)
+            }
     }
 
     @Test
@@ -300,7 +293,7 @@ class SourceSetConfigurationsTest(val useKSP2: Boolean) {
             it.startsWith("kapt") && !it.startsWith("kaptClasspath_")
         }
         val kspConfigurations = configurations.filter {
-            it.startsWith("ksp") && !it.endsWith("KotlinProcessorClasspath")
+            it.startsWith("ksp") && !it.endsWith("KotlinProcessorClasspath") && !it.startsWith("kspPluginClasspath")
         }
         assertThat(kspConfigurations).containsExactlyElementsIn(
             kaptConfigurations.map {

